### PR TITLE
Extract Levels

### PR DESCRIPTION
**Pull request by mkeshavarzi**
Extract levels command line tool with tests.

### CODE REVIEW

**Comment by pablo-mayrgundter** [comment]
Remember to request me as a reviewer or I won't see it in my queue.  I just did it, but looks like this before:

image

File: lib/src/extractLevels.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
+import fs from 'fs'
+import IfcModel from './IfcModel.js'
+
+
+const IFCSpacesELementsArrAll = []
```
Comment (pablo-mayrgundter) [inline lib/src/extractLevels.js:5]: How about IfcSpaceElements and move out of the global context.

File: lib/src/extractLevels.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+
+/**
+ * Group related element arrays for removal of each level
+ * @param {array} elementArray Extracted related element arrays from IFC file
+ * @return {array} groupedElements
+ */
+function groupElements(elementArray) {
+  const groupedElements = []
+  for (let i = 0; i<elementArray.length+1; i++) {
```
Comment (pablo-mayrgundter) [inline lib/src/extractLevels.js:80]: are you not seeing lint warnings about spacing here?
```
  for (let i = 0; i < elementArray.length + 1; i++) {
```

File: lib/src/extractLevels.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+ * @return {Map} Map with Express IDs as keys
+ */
+function expressIDtoLineMap(IFCdata) {
+  const expressIDMaps = new Map()
+  const lines = IFCdata.split('\n')
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i]
+    {
+      if ((line.indexOf('#') != -1) && (line.indexOf('=') != -1)) {
+        const data1 = line.split('#')
```
Comment (pablo-mayrgundter) [inline lib/src/extractLevels.js:117]: You sure this is right?  hashes both define elements and are used for refs, e.g.:
```
#5=IFCAPPLICATION(#1,'2021','Autodesk Revit 2021 (ENU)','Revit');
```

File: lib/src/extractLevels.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+/** @return {string} filename from command line. */
+export function getArgInputFilename() {
+  const args = process.argv.slice(2)
+  if (args.length < 1) {
+    console.error(getUsage())
+    return
+  }
+  const ifcFilename = args[0]
+  console.log('Input file: ', ifcFilename)
```
Comment (pablo-mayrgundter) [inline lib/src/extractLevels.js:37]: please add:
```
TODO(pablo): switch to logger
```

File: lib/src/extractLevels.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+ * @param {string} filename - filename of new IFC file to remove elements
+ * @param {array} expressIDsToRemove - array of expressIDs to remove
+ */
+function removeLinesFromIFCwithExpressID(filename, expressIDsToRemove) {
+  const fileCopiedBool = true
+  if (fileCopiedBool) {
+    fs.readFile(filename, 'utf8', (err, data) => {
+      if (err) throw err
+      const linesToRemove = findLinesWithExpressID(data, expressIDsToRemove)
+      // console.log(`Lines to remove from ${filename} are ${linesToRemove}`)
```
Comment (pablo-mayrgundter) [inline lib/src/extractLevels.js:169]: remove debug comments, here and throughout

File: lib/src/floodFilling.js [hunk 6]
```diff
@@ -0,0 +1,352 @@
+/* eslint-disable */
```
Comment (pablo-mayrgundter) [inline lib/src/floodFilling.js:1]: remove

File: lib/src/floodFilling.js [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+let FFAttempt = 0
+
+const FFmaxAttempt = 20
+const voxelSize = 1
+const maxThreshold = 10000
+
+
+export const boundingBoxesAll = [box1, box2, box3, box4, box5, box6, box7, box8, box9]
+
+RecursiveFloodFilling()
```
Comment (pablo-mayrgundter) [inline lib/src/floodFilling.js:47]: Move this top-level call out to a small main-fill.js or smth file.  We'll integrate it with the tool's main.js soon.  That way this file can be tested without having its code run on import.

File: lib/src/floodFilling.js [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+  SatisfyFirstRandom(pointsAccepted, pointsVisited, voxelSize, IFCUnionBox, boundingBoxesAll)
+
+  pointsLeft ++
+
+  let intVis = 0
+  while ((pointsLeft>0)&&(intVis<maxThreshold)) {
+    try {
+      CheckVoxelSurrounding(pointsAccepted[intVis], voxelSize, IFCUnionBox, boundingBoxesAll)
+    } catch {
+      console.log('Error was found on pAccepted '+intVis)
```
Comment (pablo-mayrgundter) [inline lib/src/floodFilling.js:60]: catch and log should be at least console.warn or perhaps console.error

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]
Also, please remove .DS_Store and package-lock.json from the PR.

You should add .DS_Store to the repo's .gitignore if it's not there.

**Comment by pablo-mayrgundter** [comment]
Yay!  Glad to finally get this checked in.  Looking good!

File: lib/src/extractLevels.js (hunk 3 above)
Comment (mkeshavarzi) [reply lib/src/extractLevels.js:117]: Yes. The next line splits uing "=" to make sure its not targeting the refs

File: lib/src/extractLevels.js (hunk 1 above)
Comment (mkeshavarzi) [reply lib/src/extractLevels.js:5]: I deleted ifcSpaceElements for now as we won't need it.

File: lib/src/extractLevels.js (hunk 5 above)
Comment (mkeshavarzi) [reply lib/src/extractLevels.js:169]: Removed console.log that were commented out

File: lib/src/floodFilling.js (hunk 6 above)
Comment (mkeshavarzi) [reply lib/src/floodFilling.js:1]: This would be in the FloodFilling PR

File: lib/src/floodFilling.js (hunk 7 above)
Comment (mkeshavarzi) [reply lib/src/floodFilling.js:47]: This would be in the FloodFilling PR

File: lib/src/floodFilling.js (hunk 8 above)
Comment (mkeshavarzi) [reply lib/src/floodFilling.js:60]: This would be in the FloodFilling PR

File: lib/src/floodFilling.js (hunk 6 above)
Comment (pablo-mayrgundter) [reply lib/src/floodFilling.js:1]: Ah, then remove from this PR, using ```git restore --staged <filename>``` for any files that are for the other PR

**Comment by pablo-mayrgundter** [comment]
Ok, looks like you have files from 2 PRs in here.  Just remove them from this one (using git restore --staged).. that will leave them in your working directory but we can submit this first.

**Comment by mkeshavarzi** [comment]
Done.

**Review by pablo-mayrgundter — APPROVED** [review]

**Pull request merged**